{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "47f9f885",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import libaries\n",
    "\n",
    "import pyBigWig\n",
    "import scipy.io as sio\n",
    "import numpy as np\n",
    "import math \n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import pandas as pd\n",
    "from heapq import nlargest\n",
    "import copy\n",
    "import matplotlib.gridspec as gridspec\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "from config_and_print import methy_directory, filtered_list, chrom_file, resolutions, output_directory\n",
    "\n",
    "# Ensure resolutions is treated as a tuple or list of strings\n",
    "if isinstance(resolutions, str):\n",
    "    resolutions = (resolutions,)\n",
    "\n",
    "# Print resolutions for debugging\n",
    "print(f\"Resolutions from config: {resolutions}\")\n",
    "\n",
    "# Extract resolution value and label from the resolutions string\n",
    "resolution_str = resolutions[0]\n",
    "\n",
    "# Debug print to check the value of resolution_str\n",
    "print(f\"Extracted resolution string: {resolution_str}\")\n",
    "\n",
    "\n",
    "#set path and resolution\n",
    "path = '/home/dwk681/workspace/Genomic_Files/BigWigFiles/hg19/'\n",
    "resolution = 1_000_000  \n",
    "\n",
    "# Define the path file with prefixes and colors in the following form\n",
    "#1       sc1.ACTTGA      red\n",
    "#2       sc1.GCCAAT      red\n",
    "#3       sc1.TAGCTT      red\n",
    "#4       sc10.TAGCTT     blue\n",
    "#\n",
    "filename = '/home/dwk681/workspace/multi_omics_hic_clustering/scNOMeHiC_20210127/name.order.HCG_methy.with_color.txt'\n",
    "\n",
    "# Initialize an empty dictionary to store cell ID and color\n",
    "cell_color_dict = {}\n",
    "\n",
    "# Open and read the file\n",
    "with open(filename, 'r') as file:\n",
    "    for line in file:\n",
    "        # Split the line into parts\n",
    "        parts = line.strip().split()\n",
    "        # Extract cell ID and color\n",
    "        cell_id = parts[1]\n",
    "        color = parts[2]\n",
    "        # Store in dictionary\n",
    "        cell_color_dict[cell_id] = color\n",
    "\n",
    "# Define the path to the tensor sample order file\n",
    "#This file contains the prefixes in the form\n",
    "#sc11.ACTTGA\n",
    "#sc11.CGATGT\n",
    "#sc11.GCCAAT\n",
    "#sc11.TAGCTT\n",
    "#sc12.ACTTGA\n",
    "#\n",
    "tensor_order_filename = '/home/dwk681/workspace/hypermatrix_test/hypermatrix/projects/single_cell_files/filtered_bam_list.txt'\n",
    "\n",
    "# Initialize a list to store the 1s and 0s\n",
    "color_vector = []\n",
    "\n",
    "# Open and read the tensor sample order file\n",
    "with open(tensor_order_filename, 'r') as file:\n",
    "    for line in file:\n",
    "        sample_id = line.strip()  # Remove any trailing newlines or spaces\n",
    "        if sample_id in cell_color_dict and cell_color_dict[sample_id] == 'red':\n",
    "            color_vector.append(1)\n",
    "        else:\n",
    "            color_vector.append(0)\n",
    "\n",
    "# Output the color vector to check\n",
    "print(len(color_vector))\n",
    "\n",
    "# Create a mapping dictionary\n",
    "color_mapping = {\n",
    "    'red': 'imr90',\n",
    "    'blue': 'gm12878'\n",
    "}\n",
    "\n",
    "# Update the dictionary using the mapping\n",
    "updated_cell_color_dict = {key: color_mapping[value] for key, value in cell_color_dict.items()}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6d627555",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "129\n"
     ]
    }
   ],
   "source": [
    "#a mappability experiment is an in silico experiment that reads throught a reference genome and calculates \n",
    "#how unique each section of the genome is. The more unique it is, the higher mappability score it acheives.\n",
    "\n",
    "bigwig_file = path + \"mappability_hg19_wgEncodeCrgMapabilityAlign36mer.bigWig\"\n",
    "\n",
    "# Open the BigWig file\n",
    "bw = pyBigWig.open(bigwig_file)\n",
    "\n",
    "# Define the chromosomes you want to analyze\n",
    "chromosomes = ['chr' + str(i) for i in range(1, 23)] \n",
    "\n",
    "# Define the threshold for removing bins based on average mappability\n",
    "threshold = 0.1  \n",
    "\n",
    "# Create a dictionary to store the bin indices to remove for each chromosome\n",
    "bins_to_remove = {}\n",
    "\n",
    "# Loop through each chromosome\n",
    "for chrom in chromosomes:\n",
    "    chrom_size = bw.chroms(chrom)\n",
    "\n",
    "    if chrom_size is None:\n",
    "        print(f\"Chromosome {chrom} not found in the BigWig file.\")\n",
    "        continue\n",
    "\n",
    "    # Calculate the number of bins based on the specified resolution\n",
    "    num_bins = math.ceil(chrom_size / resolution) #last bin may not be of size resolution\n",
    "\n",
    "    # Create lists to store bin indices to remove\n",
    "    remove_indices = []\n",
    "\n",
    "    # Calculate average mappability for each bin\n",
    "    for i in range(num_bins):\n",
    "        # Determine the start and end positions of the bin\n",
    "        start = i * resolution\n",
    "        end = min((i + 1) * resolution, chrom_size)  # to account for last bin which may be incomplete\n",
    "\n",
    "        # Extract the mappability values for the bin\n",
    "        values = np.nan_to_num(bw.values(chrom, start, end))\n",
    "\n",
    "        # Calculate the average mappability score for the bin\n",
    "        avg_mappability = np.mean(values)\n",
    "\n",
    "        # Check if the average mappability is below the threshold\n",
    "        if avg_mappability < threshold:\n",
    "            remove_indices.append(i)\n",
    "\n",
    "    # Store the bin indices to remove for this chromosome\n",
    "    bins_to_remove[chrom] = remove_indices\n",
    "\n",
    "# Close the BigWig file\n",
    "bw.close()\n",
    "\n",
    "# Convert the lists in bins_to_remove to numpy arrays\n",
    "for chrom in bins_to_remove:\n",
    "    bins_to_remove[chrom] = np.array(bins_to_remove[chrom])\n",
    "\n",
    "# Save the dictionary as an .npz file\n",
    "np.savez(f'/home/dwk681/workspace/multi_omics_hic_clustering/scNOMeHiC_20210127/250Kb/tensor_250Kb_dir/bins_to_remove_res{resolution}.npz', **bins_to_remove)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64eeeabb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiomics6",
   "language": "python",
   "name": "multiomics6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
