{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "76a79fd8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Resolutions from config: ('1000000:1Mb',)\n",
      "Extracted resolution string: 1000000:1Mb\n",
      "bulk_data loaded from ../../projects/single_cell_files/nan_removed_bulk_data.pkl\n",
      "original_bulk_data loaded from ../../projects/single_cell_files/nan_removed_bulk_data.pkl\n",
      "chromosome_results loaded from ../../projects/single_cell_files/chromosome_results.pkl\n",
      "Genome-wide correlation between IMR90 and GM12878: 0.6687364447778326\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "from config_and_print import methy_directory, filtered_list, chrom_file, resolutions, output_directory, mappability_threshold, normalization\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "# Ensure resolutions is treated as a tuple or list of strings\n",
    "if isinstance(resolutions, str):\n",
    "    resolutions = (resolutions,)\n",
    "\n",
    "# Print resolutions for debugging \n",
    "print(f\"Resolutions from config: {resolutions}\")\n",
    "\n",
    "# Extract resolution value and label from the resolutions string\n",
    "resolution_str = resolutions[0]\n",
    "\n",
    "# Debug print to check the value of resolution_str\n",
    "print(f\"Extracted resolution string: {resolution_str}\")\n",
    "\n",
    "def parse_resolution(resolution_str):\n",
    "    if ':' in resolution_str:\n",
    "        resolution_value, resolution_label = resolution_str.split(':')\n",
    "        try:\n",
    "            resolution = int(resolution_value)\n",
    "            return resolution, resolution_label\n",
    "        except ValueError:\n",
    "            raise ValueError(f\"Resolution value should be an integer: '{resolution_value}' in '{resolution_str}'\")\n",
    "    else:\n",
    "        raise ValueError(f\"Invalid resolution format: '{resolution_str}'. Expected format 'value:label', e.g., '1000000:1Mb'.\")\n",
    "\n",
    "resolution, resolution_label = parse_resolution(resolution_str)\n",
    "\n",
    "# Load original_bulk_data from the file\n",
    "bulk_data_input_file = '../../projects/single_cell_files/nan_removed_bulk_data.pkl'\n",
    "with open(bulk_data_input_file, 'rb') as f:\n",
    "    bulk_data = pickle.load(f)\n",
    "print(f\"bulk_data loaded from {bulk_data_input_file}\")\n",
    "\n",
    "# Load original_bulk_data from the file\n",
    "original_bulk_data_input_file = '../../projects/single_cell_files/original_bulk_data.pkl'\n",
    "with open(original_bulk_data_input_file, 'rb') as f:\n",
    "    original_bulk_data = pickle.load(f)\n",
    "print(f\"original_bulk_data loaded from {bulk_data_input_file}\")\n",
    "\n",
    "# Load chromosome_results from the file\n",
    "chromosome_results_input_file = '../../projects/single_cell_files/chromosome_results.pkl'\n",
    "with open(chromosome_results_input_file, 'rb') as f:\n",
    "    chromosome_results = pickle.load(f)\n",
    "print(f\"chromosome_results loaded from {chromosome_results_input_file}\")\n",
    "\n",
    "########################################################################\n",
    "#compute the genome-wide correlation of the two AB compartment cell types\n",
    "# should be somehting like .6 if all the above steps were done correctly\n",
    "# this is a sanity check\n",
    "# There seems to be a problem here\n",
    "# I beleive the issue is that without KR normalization, \n",
    "# the genome wide correlation of bulk is not that high\n",
    "########################################################################\n",
    "# Initialize lists to collect the eigenvector values for IMR90 and GM12878\n",
    "imr90_values = []\n",
    "gm12878_values = []\n",
    "\n",
    "# Loop through each chromosome\n",
    "for i in range(1, 23):\n",
    "    key_gm12878 = f'res{resolution}_ch{i}_oe_GM12878_{normalization}_eigenvector'\n",
    "    key_imr90 = f'res{resolution}_ch{i}_oe_IMR90_{normalization}_eigenvector'                                                               \n",
    "                                                               \n",
    "    # Check if the keys exist in the data to avoid KeyErrors\n",
    "    if key_gm12878 in bulk_data and key_imr90 in bulk_data:\n",
    "        # Extend the list with the eigenvector values from each chromosome\n",
    "        # Ensure values are numeric (floats or ints)\n",
    "        gm12878_vals = bulk_data[key_gm12878]['eigenvalue']\n",
    "        imr90_vals = bulk_data[key_imr90]['eigenvalue']\n",
    "        imr90_values.extend(imr90_vals)\n",
    "        gm12878_values.extend(gm12878_vals)\n",
    "        \n",
    "# Convert lists to pandas Series for correlation calculation\n",
    "imr90_series = pd.Series(imr90_values)\n",
    "gm12878_series = pd.Series(gm12878_values)\n",
    "\n",
    "# Calculate the Pearson correlation coefficient between IMR90 and GM12878\n",
    "correlation = imr90_series.corr(gm12878_series)\n",
    "\n",
    "print(f'Genome-wide correlation between IMR90 and GM12878: {correlation}')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "98e3fcd6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>A/B Compartment</th>\n",
       "      <th>Correlation With Bulk</th>\n",
       "      <th>Correlation With Other Bulk</th>\n",
       "      <th>Cell Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sc11.ACTTGA</td>\n",
       "      <td>[-0.001996260134358732, 0.0029843310069772146,...</td>\n",
       "      <td>0.017911</td>\n",
       "      <td>0.017911</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sc11.CGATGT</td>\n",
       "      <td>[-0.002208453420919976, 0.003204905183222053, ...</td>\n",
       "      <td>0.089631</td>\n",
       "      <td>0.099183</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sc11.GCCAAT</td>\n",
       "      <td>[0.002464560099677278, -0.0017747919120559667,...</td>\n",
       "      <td>0.122899</td>\n",
       "      <td>0.115505</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sc11.TAGCTT</td>\n",
       "      <td>[-0.009971582108582061, -0.002887513219909002,...</td>\n",
       "      <td>0.037295</td>\n",
       "      <td>0.037295</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sc12.ACTTGA</td>\n",
       "      <td>[-0.009988571147182502, -0.009985209305810186,...</td>\n",
       "      <td>0.028061</td>\n",
       "      <td>0.017026</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>sc8.ACTTGA</td>\n",
       "      <td>[-0.0011801374463571383, -0.003539522543757835...</td>\n",
       "      <td>0.107674</td>\n",
       "      <td>0.016432</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>sc9.ACTTGA</td>\n",
       "      <td>[0.004957549130336374, 0.003765320263677944, -...</td>\n",
       "      <td>0.050184</td>\n",
       "      <td>0.050184</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>sc9.CGATGT</td>\n",
       "      <td>[-0.009055192038077956, -0.0047701067079315915...</td>\n",
       "      <td>0.072359</td>\n",
       "      <td>0.037329</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>sc9.GCCAAT</td>\n",
       "      <td>[-0.0003695266934496972, -0.004470923380794043...</td>\n",
       "      <td>0.119076</td>\n",
       "      <td>0.119076</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>sc9.TAGCTT</td>\n",
       "      <td>[-0.0032956287072831384, 0.009999851390800143,...</td>\n",
       "      <td>0.054835</td>\n",
       "      <td>0.054835</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample                                    A/B Compartment  \\\n",
       "0    sc11.ACTTGA  [-0.001996260134358732, 0.0029843310069772146,...   \n",
       "1    sc11.CGATGT  [-0.002208453420919976, 0.003204905183222053, ...   \n",
       "2    sc11.GCCAAT  [0.002464560099677278, -0.0017747919120559667,...   \n",
       "3    sc11.TAGCTT  [-0.009971582108582061, -0.002887513219909002,...   \n",
       "4    sc12.ACTTGA  [-0.009988571147182502, -0.009985209305810186,...   \n",
       "..           ...                                                ...   \n",
       "124   sc8.ACTTGA  [-0.0011801374463571383, -0.003539522543757835...   \n",
       "125   sc9.ACTTGA  [0.004957549130336374, 0.003765320263677944, -...   \n",
       "126   sc9.CGATGT  [-0.009055192038077956, -0.0047701067079315915...   \n",
       "127   sc9.GCCAAT  [-0.0003695266934496972, -0.004470923380794043...   \n",
       "128   sc9.TAGCTT  [-0.0032956287072831384, 0.009999851390800143,...   \n",
       "\n",
       "     Correlation With Bulk  Correlation With Other Bulk Cell Type  \n",
       "0                 0.017911                     0.017911   gm12878  \n",
       "1                 0.089631                     0.099183     imr90  \n",
       "2                 0.122899                     0.115505     imr90  \n",
       "3                 0.037295                     0.037295   gm12878  \n",
       "4                 0.028061                     0.017026     imr90  \n",
       "..                     ...                          ...       ...  \n",
       "124               0.107674                     0.016432     imr90  \n",
       "125               0.050184                     0.050184   gm12878  \n",
       "126               0.072359                     0.037329     imr90  \n",
       "127               0.119076                     0.119076   gm12878  \n",
       "128               0.054835                     0.054835   gm12878  \n",
       "\n",
       "[129 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosome_results['chr11']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c259b22c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8e1a72d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "35822d6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average correlations for GM12878 with bulk GM12878 and bulk IMR90 by chromosome:\n",
      "Chromosome 1: 0.0700, 0.0700\n",
      "Chromosome 2: 0.0646, 0.0646\n",
      "Chromosome 3: 0.0810, 0.0810\n",
      "Chromosome 4: 0.0654, 0.0654\n",
      "Chromosome 5: 0.0942, 0.0942\n",
      "Chromosome 6: 0.0749, 0.0749\n",
      "Chromosome 7: 0.0752, 0.0752\n",
      "Chromosome 8: 0.0741, 0.0741\n",
      "Chromosome 9: 0.0965, 0.0965\n",
      "Chromosome 10: 0.0880, 0.0880\n",
      "Chromosome 11: 0.0812, 0.0812\n",
      "Chromosome 12: 0.0809, 0.0809\n",
      "Chromosome 13: 0.1220, 0.1220\n",
      "Chromosome 14: 0.1043, 0.1043\n",
      "Chromosome 15: 0.1871, 0.1871\n",
      "Chromosome 16: 0.1141, 0.1141\n",
      "Chromosome 17: 0.1089, 0.1089\n",
      "Chromosome 18: 0.1112, 0.1112\n",
      "Chromosome 19: 0.1236, 0.1236\n",
      "Chromosome 20: 0.1301, 0.1301\n",
      "Chromosome 21: 0.2080, 0.2080\n",
      "Chromosome 22: 0.1421, 0.1421\n",
      "\n",
      "Average correlations for IMR90 with bulk IMR90 and bulk GM12878 by chromosome:\n",
      "Chromosome 1: 0.0617, nan\n",
      "Chromosome 2: 0.0629, nan\n",
      "Chromosome 3: 0.0678, nan\n",
      "Chromosome 4: 0.0714, nan\n",
      "Chromosome 5: 0.0678, nan\n",
      "Chromosome 6: 0.0820, nan\n",
      "Chromosome 7: 0.0763, nan\n",
      "Chromosome 8: 0.0797, nan\n",
      "Chromosome 9: 0.0851, nan\n",
      "Chromosome 10: 0.0828, nan\n",
      "Chromosome 11: 0.0811, nan\n",
      "Chromosome 12: 0.0826, nan\n",
      "Chromosome 13: 0.1032, nan\n",
      "Chromosome 14: 0.1041, nan\n",
      "Chromosome 15: 0.1683, nan\n",
      "Chromosome 16: 0.1195, nan\n",
      "Chromosome 17: 0.1047, nan\n",
      "Chromosome 18: 0.1177, nan\n",
      "Chromosome 19: 0.1357, nan\n",
      "Chromosome 20: 0.1127, nan\n",
      "Chromosome 21: 0.2460, nan\n",
      "Chromosome 22: 0.1770, 0.1308\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
      "/tmp/ipykernel_140390/4251229429.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n"
     ]
    }
   ],
   "source": [
    "def calculate_average_correlations_from_precomputed(chromosome_results, chromosomes):\n",
    "    gm12878_avg_correlations = {}\n",
    "    imr90_avg_correlations = {}\n",
    "    gm12878_avg_correlations_withIMR90 = {}\n",
    "    imr90_avg_correlations_withGM12878 = {}\n",
    "    \n",
    "    for chr in chromosomes:\n",
    "        # Access the specific chromosome data\n",
    "        sc_dataframe = chromosome_results[f'chr{chr}']\n",
    "\n",
    "        # Extract correlations for GM12878\n",
    "        gm12878_correlations = sc_dataframe[sc_dataframe['Cell Type'] == 'gm12878']['Correlation With Bulk']\n",
    "        gm12878_avg_correlations[chr] = gm12878_correlations.mean()\n",
    "        gm12878_correlations_withIMR90 = sc_dataframe[sc_dataframe['Cell Type'] == 'gm12878']['Correlation With Other Bulk']\n",
    "        gm12878_avg_correlations_withIMR90[chr] = gm12878_correlations_withIMR90.mean()\n",
    "        \n",
    "        # Extract correlations for IMR90\n",
    "        imr90_correlations = sc_dataframe[sc_dataframe['Cell Type'] == 'imr90']['Correlation With Bulk']\n",
    "        imr90_avg_correlations[chr] = imr90_correlations.mean()\n",
    "        imr90_avg_correlations_withGM12878 = sc_dataframe[sc_dataframe['Cell Type'] == 'imr90']['Correlation With Other Bulk']\n",
    "        imr90_avg_correlations_withGM12878[chr] = imr90_avg_correlations_withGM12878.mean()\n",
    "\n",
    "    return gm12878_avg_correlations, imr90_avg_correlations, gm12878_avg_correlations_withIMR90, imr90_avg_correlations_withGM12878\n",
    "\n",
    "# Example usage\n",
    "chromosomes = [str(i) for i in range(1, 23)] \n",
    "gm12878_avg_corr, imr90_avg_corr, gm12878_avg_correlations_withIMR90, imr90_avg_correlations_withGM12878 = calculate_average_correlations_from_precomputed(chromosome_results, chromosomes)\n",
    "\n",
    "# Print the results\n",
    "print(\"Average correlations for GM12878 with bulk GM12878 and bulk IMR90 by chromosome:\")\n",
    "for chr in chromosomes:\n",
    "    corr = gm12878_avg_corr.get(chr, np.nan)\n",
    "    corr_other = gm12878_avg_correlations_withIMR90.get(chr, np.nan)\n",
    "    print(f\"Chromosome {chr}: {corr:.4f}, {corr_other:.4f}\")\n",
    "\n",
    "print(\"\\nAverage correlations for IMR90 with bulk IMR90 and bulk GM12878 by chromosome:\")\n",
    "for chr in chromosomes:\n",
    "    corr = imr90_avg_corr.get(chr, np.nan)\n",
    "    corr_other = imr90_avg_correlations_withGM12878.get(chr, np.nan)\n",
    "    print(f\"Chromosome {chr}: {corr:.4f}, {corr_other:.4f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d47eb16d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "25b0bc1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>A/B Compartment</th>\n",
       "      <th>Correlation With Bulk</th>\n",
       "      <th>Correlation With Other Bulk</th>\n",
       "      <th>Cell Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sc11.ACTTGA</td>\n",
       "      <td>[nan, 0.00999999999887582, 0.00999999999991210...</td>\n",
       "      <td>0.033589</td>\n",
       "      <td>0.033589</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sc11.CGATGT</td>\n",
       "      <td>[nan, -0.00860145974174996, -0.008085117258960...</td>\n",
       "      <td>0.013052</td>\n",
       "      <td>-0.004874</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sc11.GCCAAT</td>\n",
       "      <td>[nan, 0.007224265287621554, 0.0021975034515107...</td>\n",
       "      <td>0.020947</td>\n",
       "      <td>0.041650</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sc11.TAGCTT</td>\n",
       "      <td>[nan, 0.00717223226846834, 0.0078593022772871,...</td>\n",
       "      <td>0.135427</td>\n",
       "      <td>0.135427</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sc12.ACTTGA</td>\n",
       "      <td>[nan, -0.006025251715588904, -0.01, 0.00387379...</td>\n",
       "      <td>0.021850</td>\n",
       "      <td>0.045165</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>sc8.ACTTGA</td>\n",
       "      <td>[nan, 0.004280592728428014, 0.0025654303469549...</td>\n",
       "      <td>0.069955</td>\n",
       "      <td>-0.113859</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>sc9.ACTTGA</td>\n",
       "      <td>[nan, 0.0016417632617453867, 0.005036153409245...</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>sc9.CGATGT</td>\n",
       "      <td>[nan, -0.004515304166640753, 0.006761261958069...</td>\n",
       "      <td>0.019136</td>\n",
       "      <td>-0.047854</td>\n",
       "      <td>imr90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>sc9.GCCAAT</td>\n",
       "      <td>[nan, 0.01, -0.0006984483014165713, -0.0021366...</td>\n",
       "      <td>0.077734</td>\n",
       "      <td>0.077734</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>sc9.TAGCTT</td>\n",
       "      <td>[nan, -0.006231309906028834, 0.002381052787404...</td>\n",
       "      <td>0.112570</td>\n",
       "      <td>0.112570</td>\n",
       "      <td>gm12878</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample                                    A/B Compartment  \\\n",
       "0    sc11.ACTTGA  [nan, 0.00999999999887582, 0.00999999999991210...   \n",
       "1    sc11.CGATGT  [nan, -0.00860145974174996, -0.008085117258960...   \n",
       "2    sc11.GCCAAT  [nan, 0.007224265287621554, 0.0021975034515107...   \n",
       "3    sc11.TAGCTT  [nan, 0.00717223226846834, 0.0078593022772871,...   \n",
       "4    sc12.ACTTGA  [nan, -0.006025251715588904, -0.01, 0.00387379...   \n",
       "..           ...                                                ...   \n",
       "124   sc8.ACTTGA  [nan, 0.004280592728428014, 0.0025654303469549...   \n",
       "125   sc9.ACTTGA  [nan, 0.0016417632617453867, 0.005036153409245...   \n",
       "126   sc9.CGATGT  [nan, -0.004515304166640753, 0.006761261958069...   \n",
       "127   sc9.GCCAAT  [nan, 0.01, -0.0006984483014165713, -0.0021366...   \n",
       "128   sc9.TAGCTT  [nan, -0.006231309906028834, 0.002381052787404...   \n",
       "\n",
       "     Correlation With Bulk  Correlation With Other Bulk Cell Type  \n",
       "0                 0.033589                     0.033589   gm12878  \n",
       "1                 0.013052                    -0.004874     imr90  \n",
       "2                 0.020947                     0.041650     imr90  \n",
       "3                 0.135427                     0.135427   gm12878  \n",
       "4                 0.021850                     0.045165     imr90  \n",
       "..                     ...                          ...       ...  \n",
       "124               0.069955                    -0.113859     imr90  \n",
       "125               0.007107                     0.007107   gm12878  \n",
       "126               0.019136                    -0.047854     imr90  \n",
       "127               0.077734                     0.077734   gm12878  \n",
       "128               0.112570                     0.112570   gm12878  \n",
       "\n",
       "[129 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr = 1\n",
    "chromosome_results[f'chr{chr}']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6f5ac12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad45c861",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr = 15\n",
    "sc_dataframe = chromosome_results[f'chr{chr}']\n",
    "\n",
    "# Define the bulk eigenvectors\n",
    "bulk_eigenvector_gm12878 = original_bulk_data[f'res1000000_ch{chr}_oe_GM12878_{normalization}_eigenvector']['eigenvalue'].values\n",
    "bulk_eigenvector_imr90 = original_bulk_data[f'res1000000_ch{chr}_oe_IMR90_{normalization}_eigenvector']['eigenvalue'].values\n",
    "\n",
    "# Sort and prepare the dataframe\n",
    "sc_dataframe['Cell Type Sort'] = sc_dataframe['Cell Type'].map({'gm12878': 0, 'imr90': 1})\n",
    "sc_dataframe = sc_dataframe.sort_values(by=['Cell Type Sort', 'Correlation With Bulk'], ascending=[True, False])\n",
    "sc_dataframe.drop('Cell Type Sort', axis=1, inplace=True)\n",
    "\n",
    "def plot_eigenvectors_with_heatmap(bulk_eigenvector1, bulk_eigenvector2, sc_dataframe, title1, title2, title, correlation, chr, save_path=None, separation_rows=5):\n",
    "    fig = plt.figure(figsize=(15, 22))  # Adjusted the figure size for better proportions\n",
    "    gs = gridspec.GridSpec(nrows=3, ncols=2, figure=fig, width_ratios=[30, 1], height_ratios=[1, 1, 4], hspace=0.3, wspace=0.05)\n",
    "\n",
    "    ax_bulk1 = fig.add_subplot(gs[0, 0])\n",
    "    ax_bulk1.plot(bulk_eigenvector1, color='orange')\n",
    "    ax_bulk1.set_title(title1, fontsize=25)\n",
    "    ax_bulk1.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_bulk1.set_ylabel(\"Eigenvector Value\", fontsize=25)\n",
    "\n",
    "    ax_bulk2 = fig.add_subplot(gs[1, 0])\n",
    "    ax_bulk2.plot(bulk_eigenvector2, color='green')\n",
    "    ax_bulk2.set_title(title2, fontsize=25)\n",
    "    ax_bulk2.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_bulk2.set_ylabel(\"Eigenvector Value\", fontsize=25)\n",
    "\n",
    "    gm12878_data = np.stack(sc_dataframe[sc_dataframe['Cell Type'] == 'gm12878']['A/B Compartment'].apply(lambda x: np.array(x)).tolist())\n",
    "    imr90_data = np.stack(sc_dataframe[sc_dataframe['Cell Type'] == 'imr90']['A/B Compartment'].apply(lambda x: np.array(x)).tolist())\n",
    "    \n",
    "    empty_rows = np.full((separation_rows, gm12878_data.shape[1]), np.nan)\n",
    "    sc_data_array = np.vstack([gm12878_data, empty_rows, imr90_data])\n",
    "\n",
    "    # Create a colormap that does not include white for NaN values\n",
    "    cmap = plt.cm.coolwarm\n",
    "    cmap.set_bad(color='black')  # Set NaN color to black\n",
    "\n",
    "    ax_sc = fig.add_subplot(gs[2, 0])\n",
    "    im = ax_sc.imshow(sc_data_array, aspect='auto', cmap=cmap, interpolation='nearest')  # Use the adjusted colormap\n",
    "    ax_sc.set_title(f\"Heatmap of Single Cell Samples for chr {chr}\", fontsize=25)  # Slightly larger font for the title\n",
    "    ax_sc.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_sc.set_ylabel(\"Sample Index\", fontsize=20)\n",
    "\n",
    "    num_gm12878 = len(gm12878_data)\n",
    "    num_imr90 = len(imr90_data)\n",
    "    \n",
    "    ax_sc.set_yticks([num_gm12878 / 2, num_gm12878 + separation_rows + num_imr90 / 2])\n",
    "    ax_sc.set_yticklabels(['GM12878', 'IMR90'], fontsize=30)\n",
    "\n",
    "    ax_cbar = fig.add_subplot(gs[2, 1])\n",
    "    fig.colorbar(im, cax=ax_cbar)\n",
    "    ax_cbar.set_ylabel('NTF Compartment Call', rotation=270, labelpad=25, fontsize=25)\n",
    "\n",
    "    plt.suptitle(title, fontsize=35)  # Increased the title font size\n",
    "\n",
    "    # Construct the caption with the chromosome number, correlation value, and cell counts\n",
    "    caption = (f\"Figure : The single cell A/B compartment calls from the NTF model visually seperate the {num_gm12878} GM12878 cells \" \n",
    "               f\"from the {num_imr90} IMR90 cells, as demonstrated on chromosome {chr} in this figure. \"\n",
    "               f\"The single cell A/B compartment calls are also visually correlated with the bulk GM12878 and bulk IMR90 Eigenvector values, \"\n",
    "               f\"as computed by the Juicer software. \"\n",
    "               f\"The genome-wide correlation between bulk GM12878 and bulk IMR90 is {correlation:.2f}, which is normal for Hi-C data \"\n",
    "               f\"with no normalization, such as KR normalization.\" )\n",
    "               \n",
    "    # Add the caption at the bottom\n",
    "    plt.figtext(0.01, -0.05, caption, wrap=True, horizontalalignment='left', fontsize=25)\n",
    "\n",
    "    if save_path:\n",
    "        # Ensure the save directory exists\n",
    "        save_directory = os.path.dirname(save_path)\n",
    "        if not os.path.exists(save_directory):\n",
    "            os.makedirs(save_directory)\n",
    "        \n",
    "        plt.savefig(save_path, bbox_inches='tight')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "# Example invocation with the genome-wide correlation value and chromosome number\n",
    "plot_eigenvectors_with_heatmap(\n",
    "    bulk_eigenvector_gm12878,\n",
    "    bulk_eigenvector_imr90,\n",
    "    sc_dataframe,\n",
    "    f\"Bulk GM12878 for chr{chr}\",\n",
    "    f\"Bulk IMR90 for chr{chr}\",\n",
    "    \"Single Cell Compartment Call Comparison\",\n",
    "    correlation=correlation,  # Use the computed correlation value here\n",
    "    chr=chr,  # Use the chromosome number variable\n",
    "    save_path=f\"../../files/AB_compartment_heatmap_ch{chr}.png\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ae5244f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiomics6",
   "language": "python",
   "name": "multiomics6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
