{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76a79fd8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bam_directory='/home/dwk681/workspace/cluster_cells_from_GSE189158_NOMe_HiC/filesFromCluster/bam'\n",
      "methy_directory='/home/dwk681/workspace/cluster_cells_from_GSE189158_NOMe_HiC/filesFromCluster/bam/methylation/filter_low_qual'\n",
      "software_directory='../../bin/softwarefiles'\n",
      "chrom_file='../../bin/softwarefiles/hg19.autosome.chrom.sizes'\n",
      "fragments_file='../../bin/softwarefiles/hg19_DpnII.txt'\n",
      "output_directory='../../projects/single_cell_files'\n",
      "hg19_fa_url='ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz'\n",
      "filtered_list='../../projects/single_cell_files/filtered_bam_list.txt'\n",
      "schicluster_env='schicluster2'\n",
      "bisulfite_env='bisulfitehic27'\n",
      "min_high_quality_reads='250000'\n",
      "resolutions='1000000:1Mb'\n",
      "impute='False'\n",
      "cluster_compartments='False'\n",
      "cumulant='False'\n",
      "iterations='400'\n",
      "chromosomes=1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22\n",
      "dark_regions_hg19_url='https://www.encodeproject.org/files/ENCFF000EHJ/@@download/ENCFF000EHJ.bigWig'\n",
      "mappability_threshold='0.6'\n",
      "data_type='oe'\n",
      "normalization='NONE'\n",
      "correlation='True'\n",
      "genomeID='hg19'\n",
      "hic_GM12878_url='https://ftp.ncbi.nlm.nih.gov/geo/series/GSE63nnn/GSE63525/suppl/GSE63525%5FGM12878%5Finsitu%5Fprimary%2Breplicate%5Fcombined%5F30%2Ehic'\n",
      "hic_IMR90_url='https://ftp.ncbi.nlm.nih.gov/geo/series/GSE63nnn/GSE63525/suppl/GSE63525%5FIMR90%5Fcombined%5F30.hic'\n",
      "Resolutions from config: ('1000000:1Mb',)\n",
      "Extracted resolution string: 1000000:1Mb\n",
      "bulk_data loaded from ../../projects/single_cell_files/nan_removed_bulk_data.pkl\n",
      "original_bulk_data loaded from ../../projects/single_cell_files/nan_removed_bulk_data.pkl\n",
      "chromosome_results loaded from ../../projects/single_cell_files/chromosome_results.pkl\n",
      "Genome-wide correlation between IMR90 and GM12878: 0.09974947590453016\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "from config_and_print import methy_directory, filtered_list, chrom_file, resolutions, output_directory, mappability_threshold, normalization\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "# Ensure resolutions is treated as a tuple or list of strings\n",
    "\n",
    "if isinstance(resolutions, str):\n",
    "    resolutions = (resolutions,)\n",
    "\n",
    "# Print resolutions for debugging \n",
    "print(f\"Resolutions from config: {resolutions}\")\n",
    "\n",
    "# Extract resolution value and label from the resolutions string\n",
    "resolution_str = resolutions[0]\n",
    "\n",
    "# Debug print to check the value of resolution_str\n",
    "print(f\"Extracted resolution string: {resolution_str}\")\n",
    "\n",
    "def parse_resolution(resolution_str):\n",
    "    if ':' in resolution_str:\n",
    "        resolution_value, resolution_label = resolution_str.split(':')\n",
    "        try:\n",
    "            resolution = int(resolution_value)\n",
    "            return resolution, resolution_label\n",
    "        except ValueError:\n",
    "            raise ValueError(f\"Resolution value should be an integer: '{resolution_value}' in '{resolution_str}'\")\n",
    "    else:\n",
    "        raise ValueError(f\"Invalid resolution format: '{resolution_str}'. Expected format 'value:label', e.g., '1000000:1Mb'.\")\n",
    "\n",
    "resolution, resolution_label = parse_resolution(resolution_str)\n",
    "\n",
    "# Load original_bulk_data from the file\n",
    "bulk_data_input_file = '../../projects/single_cell_files/nan_removed_bulk_data.pkl'\n",
    "with open(bulk_data_input_file, 'rb') as f:\n",
    "    bulk_data = pickle.load(f)\n",
    "print(f\"bulk_data loaded from {bulk_data_input_file}\")\n",
    "\n",
    "# Load original_bulk_data from the file\n",
    "original_bulk_data_input_file = '../../projects/single_cell_files/original_bulk_data.pkl'\n",
    "with open(original_bulk_data_input_file, 'rb') as f:\n",
    "    original_bulk_data = pickle.load(f)\n",
    "print(f\"original_bulk_data loaded from {bulk_data_input_file}\")\n",
    "\n",
    "# Load chromosome_results from the file\n",
    "chromosome_results_input_file = '../../projects/single_cell_files/chromosome_results.pkl'\n",
    "with open(chromosome_results_input_file, 'rb') as f:\n",
    "    chromosome_results = pickle.load(f)\n",
    "print(f\"chromosome_results loaded from {chromosome_results_input_file}\")\n",
    "\n",
    "########################################################################\n",
    "#compute the genome-wide correlation of the two AB compartment cell types\n",
    "# should be somehting like .6 if all the above steps were done correctly\n",
    "# this is a sanity check\n",
    "# There seems to be a problem here\n",
    "# I beleive the issue is that without KR normalization, \n",
    "# the genome wide correlation of bulk is not that high\n",
    "########################################################################\n",
    "# Initialize lists to collect the eigenvector values for IMR90 and GM12878\n",
    "imr90_values = []\n",
    "gm12878_values = []\n",
    "\n",
    "# Loop through each chromosome\n",
    "for i in range(1, 23):\n",
    "    key_gm12878 = f'res{resolution}_ch{i}_oe_GM12878_{normalization}_eigenvector'\n",
    "    key_imr90 = f'res{resolution}_ch{i}_oe_IMR90_{normalization}_eigenvector'                                                               \n",
    "                                                               \n",
    "    # Check if the keys exist in the data to avoid KeyErrors\n",
    "    if key_gm12878 in bulk_data and key_imr90 in bulk_data:\n",
    "        # Extend the list with the eigenvector values from each chromosome\n",
    "        # Ensure values are numeric (floats or ints)\n",
    "        gm12878_vals = bulk_data[key_gm12878]['eigenvalue']\n",
    "        imr90_vals = bulk_data[key_imr90]['eigenvalue']\n",
    "        imr90_values.extend(imr90_vals)\n",
    "        gm12878_values.extend(gm12878_vals)\n",
    "        \n",
    "# Convert lists to pandas Series for correlation calculation\n",
    "imr90_series = pd.Series(imr90_values)\n",
    "gm12878_series = pd.Series(gm12878_values)\n",
    "\n",
    "# Calculate the Pearson correlation coefficient between IMR90 and GM12878\n",
    "correlation = imr90_series.corr(gm12878_series)\n",
    "\n",
    "print(f'Genome-wide correlation between IMR90 and GM12878: {correlation}')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eced7bbd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#chromosome_results['chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "65f68cd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eigenvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.147524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.145734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.143773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.114985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.068368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>-0.140899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>-0.027149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>-0.051354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>0.024367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>-0.057279</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    eigenvalue\n",
       "0    -0.147524\n",
       "1    -0.145734\n",
       "2    -0.143773\n",
       "3    -0.114985\n",
       "4     0.068368\n",
       "..         ...\n",
       "69   -0.140899\n",
       "70   -0.027149\n",
       "71   -0.051354\n",
       "72    0.024367\n",
       "73   -0.057279\n",
       "\n",
       "[74 rows x 1 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulk_data['res1000000_ch15_oe_GM12878_NONE_eigenvector']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33015b6d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "427db9a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_average_correlations_from_precomputed(chromosome_results, chromosomes):\n",
    "    gm12878_avg_correlations = {}\n",
    "    imr90_avg_correlations = {}\n",
    "\n",
    "    for chr in chromosomes:\n",
    "        # Access the specific chromosome data\n",
    "        sc_dataframe = chromosome_results[f'chr{chr}']\n",
    "\n",
    "        # Extract correlations for GM12878\n",
    "        gm12878_correlations = sc_dataframe[sc_dataframe['Cell Type'] == 'gm12878']['Correlation With Bulk']\n",
    "        gm12878_avg_correlations[chr] = gm12878_correlations.mean()\n",
    "        \n",
    "        # Extract correlations for IMR90\n",
    "        imr90_correlations = sc_dataframe[sc_dataframe['Cell Type'] == 'imr90']['Correlation With Bulk']\n",
    "        imr90_avg_correlations[chr] = imr90_correlations.mean()\n",
    "\n",
    "    return gm12878_avg_correlations, imr90_avg_correlations\n",
    "\n",
    "# Example usage\n",
    "chromosomes = [str(i) for i in range(1, 23)] \n",
    "gm12878_avg_corr, imr90_avg_corr = calculate_average_correlations_from_precomputed(chromosome_results, chromosomes)\n",
    "\n",
    "# Print the results\n",
    "print(\"Average correlations for GM12878 with bulk GM12878 by chromosome:\")\n",
    "for chr, corr in gm12878_avg_corr.items():\n",
    "    print(f\"Chromosome {chr}: {corr:.4f}\")\n",
    "\n",
    "print(\"\\nAverage correlations for IMR90 with bulk IMR90 by chromosome:\")\n",
    "for chr, corr in imr90_avg_corr.items():\n",
    "    print(f\"Chromosome {chr}: {corr:.4f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf7eb65",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr = 1\n",
    "chromosome_results[f'chr{chr}']['Correlation With Bulk']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b78f9da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7fe2df0",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr = 1\n",
    "sc_dataframe = chromosome_results[f'chr{chr}']\n",
    "\n",
    "# Define the bulk eigenvectors\n",
    "bulk_eigenvector_gm12878 = bulk_data[f'res1000000_ch{chr}_oe_GM12878_{normalization}_eigenvector']['eigenvalue'].values\n",
    "bulk_eigenvector_imr90 = bulk_data[f'res1000000_ch{chr}_oe_IMR90_{normalization}_eigenvector']['eigenvalue'].values\n",
    "\n",
    "# Sort and prepare the dataframe\n",
    "sc_dataframe['Cell Type Sort'] = sc_dataframe['Cell Type'].map({'gm12878': 0, 'imr90': 1})\n",
    "sc_dataframe = sc_dataframe.sort_values(by=['Cell Type Sort', 'Correlation With Bulk'], ascending=[True, False])\n",
    "sc_dataframe.drop('Cell Type Sort', axis=1, inplace=True)\n",
    "\n",
    "def plot_eigenvectors_with_heatmap(bulk_eigenvector1, bulk_eigenvector2, sc_dataframe, title1, title2, title, correlation, chr, save_path=None, separation_rows=5):\n",
    "    fig = plt.figure(figsize=(15, 22))  # Adjusted the figure size for better proportions\n",
    "    gs = gridspec.GridSpec(nrows=3, ncols=2, figure=fig, width_ratios=[30, 1], height_ratios=[1, 1, 4], hspace=0.3, wspace=0.05)\n",
    "\n",
    "    ax_bulk1 = fig.add_subplot(gs[0, 0])\n",
    "    ax_bulk1.plot(bulk_eigenvector1, color='orange')\n",
    "    ax_bulk1.set_title(title1, fontsize=25)\n",
    "    ax_bulk1.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_bulk1.set_ylabel(\"Eigenvector Value\", fontsize=25)\n",
    "\n",
    "    ax_bulk2 = fig.add_subplot(gs[1, 0])\n",
    "    ax_bulk2.plot(bulk_eigenvector2, color='green')\n",
    "    ax_bulk2.set_title(title2, fontsize=25)\n",
    "    ax_bulk2.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_bulk2.set_ylabel(\"Eigenvector Value\", fontsize=25)\n",
    "\n",
    "    gm12878_data = np.stack(sc_dataframe[sc_dataframe['Cell Type'] == 'gm12878']['A/B Compartment'].apply(lambda x: np.array(x)).tolist())\n",
    "    imr90_data = np.stack(sc_dataframe[sc_dataframe['Cell Type'] == 'imr90']['A/B Compartment'].apply(lambda x: np.array(x)).tolist())\n",
    "    \n",
    "    empty_rows = np.full((separation_rows, gm12878_data.shape[1]), np.nan)\n",
    "    sc_data_array = np.vstack([gm12878_data, empty_rows, imr90_data])\n",
    "\n",
    "    # Create a colormap that does not include white for NaN values\n",
    "    cmap = plt.cm.coolwarm\n",
    "    cmap.set_bad(color='black')  # Set NaN color to black\n",
    "\n",
    "    ax_sc = fig.add_subplot(gs[2, 0])\n",
    "    im = ax_sc.imshow(sc_data_array, aspect='auto', cmap=cmap, interpolation='nearest')  # Use the adjusted colormap\n",
    "    ax_sc.set_title(f\"Heatmap of Single Cell Samples for chr {chr}\", fontsize=25)  # Slightly larger font for the title\n",
    "    ax_sc.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_sc.set_ylabel(\"Sample Index\", fontsize=20)\n",
    "\n",
    "    num_gm12878 = len(gm12878_data)\n",
    "    num_imr90 = len(imr90_data)\n",
    "    \n",
    "    ax_sc.set_yticks([num_gm12878 / 2, num_gm12878 + separation_rows + num_imr90 / 2])\n",
    "    ax_sc.set_yticklabels(['GM12878', 'IMR90'], fontsize=30)\n",
    "\n",
    "    ax_cbar = fig.add_subplot(gs[2, 1])\n",
    "    fig.colorbar(im, cax=ax_cbar)\n",
    "    ax_cbar.set_ylabel('NTF Compartment Call', rotation=270, labelpad=25, fontsize=25)\n",
    "\n",
    "    plt.suptitle(title, fontsize=35)  # Increased the title font size\n",
    "\n",
    "    # Construct the caption with the chromosome number, correlation value, and cell counts\n",
    "    caption = (f\"Figure : The single cell A/B compartment calls from the NTF model visually seperate the {num_gm12878} GM12878 cells \" \n",
    "               f\"from the {num_imr90} IMR90 cells, as demonstrated on chromosome {chr} in this figure. \"\n",
    "               f\"The single cell A/B compartment calls are also visually correlated with the bulk GM12878 and bulk IMR90 Eigenvector values, \"\n",
    "               f\"as computed by the Juicer software. \"\n",
    "               f\"The genome-wide correlation between bulk GM12878 and bulk IMR90 is {correlation:.2f}, which is normal for Hi-C data \"\n",
    "               f\"with no normalization, such as KR normalization.\" )\n",
    "               \n",
    "    # Add the caption at the bottom\n",
    "    plt.figtext(0.01, -0.05, caption, wrap=True, horizontalalignment='left', fontsize=25)\n",
    "\n",
    "    if save_path:\n",
    "        # Ensure the save directory exists\n",
    "        save_directory = os.path.dirname(save_path)\n",
    "        if not os.path.exists(save_directory):\n",
    "            os.makedirs(save_directory)\n",
    "        \n",
    "        plt.savefig(save_path, bbox_inches='tight')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "# Example invocation with the genome-wide correlation value and chromosome number\n",
    "plot_eigenvectors_with_heatmap(\n",
    "    bulk_eigenvector_gm12878,\n",
    "    bulk_eigenvector_imr90,\n",
    "    sc_dataframe,\n",
    "    f\"Bulk GM12878 for chr{chr}\",\n",
    "    f\"Bulk IMR90 for chr{chr}\",\n",
    "    \"Single Cell Compartment Call Comparison\",\n",
    "    correlation=correlation,  # Use the computed correlation value here\n",
    "    chr=chr,  # Use the chromosome number variable\n",
    "    save_path=f\"../../files/AB_compartment_heatmap_ch{chr}.png\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "744f8f52",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiomics6",
   "language": "python",
   "name": "multiomics6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
