{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "91fd9ced",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyBigWig\n",
    "import numpy as np\n",
    "import h5py\n",
    "import os\n",
    "from scipy.stats import ks_2samp\n",
    "from multiprocessing import Pool\n",
    "\n",
    "# Constants for bin sizes\n",
    "large_bin_size = 1_000_000\n",
    "small_bin_size = 10_000\n",
    "\n",
    "#large_bin_size = 250_000  # 250 kb\n",
    "#small_bin_size = 2_500   # 10 kb, gives 25 smaller bins per large bin\n",
    "\n",
    "# Chromosome sizes\n",
    "chromosomes = {\n",
    "    '1': 249250621, '2': 243199373, '3': 198022430, '4': 191154276, '5': 180915260,\n",
    "    '6': 171115067, '7': 159138663, '8': 146364022, '9': 141213431, '10': 135534747,\n",
    "    '11': 135006516, '12': 133851895, '13': 115169878, '14': 107349540, '15': 102531392,\n",
    "    '16': 90354753, '17': 81195210, '18': 78077248, '19': 59128983, '20': 63025520,\n",
    "    '21': 48129895, '22': 51304566\n",
    "}\n",
    "\n",
    "def correlation_matrix(data):\n",
    "    # Compute the correlation matrix using numpy\n",
    "    correlation_mat = np.corrcoef(data, rowvar=False)\n",
    "    return correlation_mat\n",
    "\n",
    "def calculate_integrals(args):\n",
    "    bigwigfile_path, chrom, chrom_length, large_bin_start = args\n",
    "    bw = pyBigWig.open(bigwigfile_path)  # Open the BigWig file here\n",
    "    integrals = np.zeros(int(large_bin_size / small_bin_size))\n",
    "\n",
    "    for j in range(int(large_bin_size / small_bin_size)):\n",
    "        small_bin_start = large_bin_start + j * small_bin_size\n",
    "        small_bin_end = small_bin_start + small_bin_size - 1\n",
    "        small_bin_start = max(small_bin_start, 0)\n",
    "        small_bin_end = min(small_bin_end, chrom_length - 1)\n",
    "\n",
    "        if small_bin_start < small_bin_end:\n",
    "            small_bin_data = bw.values(chrom, small_bin_start, small_bin_end)\n",
    "            integral = np.trapz(small_bin_data) if small_bin_data is not None else 0\n",
    "            integrals[j] = integral if not np.isnan(integral) else 0\n",
    "\n",
    "    bw.close()\n",
    "    return integrals\n",
    "\n",
    "def process_chrom(bigwigfile_path, chrom, file_prefix):\n",
    "    output_dir = f'/home/dwk681/workspace/multi_omics_hic_clustering/scNOMeHiC_20210127/1Mb/methylation_1Mb_p_value_ks_correlation_dir/chr{chrom}/'\n",
    "    p_value_filename = os.path.join(output_dir, f'{file_prefix}_{chrom}_ks_p_value_adjusted.h5')\n",
    "\n",
    "    # Check if the file already exists\n",
    "    if os.path.exists(p_value_filename):\n",
    "        print(f\"File {p_value_filename} already exists. Skipping.\")\n",
    "        return\n",
    "\n",
    "    chrom_length = chromosomes[chrom]\n",
    "    large_bins = range(0, chrom_length, large_bin_size)\n",
    "    num_large_bins = len(large_bins)\n",
    "    integrals = np.zeros((int(large_bin_size / small_bin_size), num_large_bins))\n",
    "\n",
    "    args_list = [(bigwigfile_path, chrom, chrom_length, lb) for lb in large_bins]\n",
    "\n",
    "    with Pool(processes=4) as pool:\n",
    "        integrals_list = pool.map(calculate_integrals, args_list)\n",
    "\n",
    "    for i, integrals_small in enumerate(integrals_list):\n",
    "        integrals[:, i] = integrals_small\n",
    "        \n",
    "    # Compute KS p-values and correlation\n",
    "    epsilon = 1e-10\n",
    "    num_rows = integrals.shape[0]\n",
    "    p_value_ks_matrix = np.zeros((num_rows, num_rows))\n",
    "\n",
    "    for i in range(num_rows):\n",
    "        for j in range(i + 1, num_rows):\n",
    "            _, ks_p_value = ks_2samp(integrals[i], integrals[j])\n",
    "            p_value_ks_matrix[i, j] = p_value_ks_matrix[j, i] = ks_p_value\n",
    "\n",
    "    corr_matrix = correlation_matrix(p_value_ks_matrix)\n",
    "    \n",
    "    # Save matrices\n",
    "    output_dir = f'/home/dwk681/workspace/multi_omics_hic_clustering/scNOMeHiC_20210127/1Mb/methylation_1Mb_p_value_ks_correlation_dir/chr{chrom}/'\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    p_value_filename = os.path.join(output_dir, f'{file_prefix}_{chrom}_ks_p_value_adjusted.h5')\n",
    "    with h5py.File(p_value_filename, 'w') as hf:\n",
    "        hf.create_dataset('matrix', data=p_value_ks_matrix)\n",
    "\n",
    "def process_directory(base_path):\n",
    "    suffix = '.methy_count.b37.bw'\n",
    "    files = [f for f in os.listdir(base_path) if f.endswith(suffix)]\n",
    "\n",
    "    for file_name in files:\n",
    "        full_path = os.path.join(base_path, file_name)\n",
    "        file_prefix = file_name[:-len(suffix)]\n",
    "        for chrom in chromosomes:\n",
    "            process_chrom(full_path, chrom, file_prefix)  # Pass the file path, not the object.\n",
    "            \n",
    "if __name__ == \"__main__\":\n",
    "    base_path = '/home/dwk681/workspace/cluster_cells_from_GSE189158_NOMe_HiC/filesFromCluster/bam/methylation/'\n",
    "    process_directory(base_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aef92c12",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiomics6",
   "language": "python",
   "name": "multiomics6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
