{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76a79fd8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bam_directory='/home/dwk681/workspace/cluster_cells_from_GSE189158_NOMe_HiC/filesFromCluster/bam'\n",
      "methy_directory='/home/dwk681/workspace/cluster_cells_from_GSE189158_NOMe_HiC/filesFromCluster/bam/methylation/filter_low_qual'\n",
      "software_directory='../../bin/softwarefiles'\n",
      "chrom_file='../../bin/softwarefiles/hg19.autosome.chrom.sizes'\n",
      "fragments_file='../../bin/softwarefiles/hg19_DpnII.txt'\n",
      "output_directory='../../projects/single_cell_files'\n",
      "hg19_fa_url='ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz'\n",
      "filtered_list='../../projects/single_cell_files/filtered_bam_list.txt'\n",
      "schicluster_env='schicluster2'\n",
      "bisulfite_env='bisulfitehic27'\n",
      "min_high_quality_reads='250000'\n",
      "resolutions='1000000:1Mb'\n",
      "impute='False'\n",
      "cluster_compartments='False'\n",
      "cumulant='False'\n",
      "iterations='400'\n",
      "chromosomes=1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22\n",
      "dark_regions_hg19_url='https://www.encodeproject.org/files/ENCFF000EHJ/@@download/ENCFF000EHJ.bigWig'\n",
      "mappability_threshold='0.6'\n",
      "data_type='oe'\n",
      "normalization='NONE'\n",
      "correlation='True'\n",
      "genomeID='hg19'\n",
      "hic_GM12878_url='https://ftp.ncbi.nlm.nih.gov/geo/series/GSE63nnn/GSE63525/suppl/GSE63525%5FGM12878%5Finsitu%5Fprimary%2Breplicate%5Fcombined%5F30%2Ehic'\n",
      "hic_IMR90_url='https://ftp.ncbi.nlm.nih.gov/geo/series/GSE63nnn/GSE63525/suppl/GSE63525%5FIMR90%5Fcombined%5F30.hic'\n",
      "original_bulk_data loaded from ../../projects/single_cell_files/nan_removed_bulk_data.pkl\n",
      "chromosome_results loaded from ../../projects/single_cell_files/chromosome_results.pkl\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "from config_and_print import methy_directory, filtered_list, chrom_file, resolutions, output_directory, mappability_threshold, normalization\n",
    "\n",
    "# Load original_bulk_data from the file\n",
    "bulk_data_input_file = '../../projects/single_cell_files/nan_removed_bulk_data.pkl'\n",
    "with open(bulk_data_input_file, 'rb') as f:\n",
    "    bulk_data = pickle.load(f)\n",
    "print(f\"original_bulk_data loaded from {bulk_data_input_file}\")\n",
    "\n",
    "# Load chromosome_results from the file\n",
    "chromosome_results_input_file = '../../projects/single_cell_files/chromosome_results.pkl'\n",
    "with open(chromosome_results_input_file, 'rb') as f:\n",
    "    chromosome_results = pickle.load(f)\n",
    "print(f\"chromosome_results loaded from {chromosome_results_input_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dec45813",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>A/B Compartment</th>\n",
       "      <th>Correlation With Bulk</th>\n",
       "      <th>Cell Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sc11.ACTTGA</td>\n",
       "      <td>[nan, 0.018999999977951314, 0.019, 0.019, 0.01...</td>\n",
       "      <td>0.115300</td>\n",
       "      <td>b'gm12878'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sc11.CGATGT</td>\n",
       "      <td>[nan, -0.001, -0.0009999976557568116, 0.015259...</td>\n",
       "      <td>0.169729</td>\n",
       "      <td>b'imr90'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sc11.GCCAAT</td>\n",
       "      <td>[nan, -0.0009980968427463922, -0.0009999999999...</td>\n",
       "      <td>0.163800</td>\n",
       "      <td>b'imr90'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sc11.TAGCTT</td>\n",
       "      <td>[nan, 0.019, 0.018717003742121565, 0.019, 0.01...</td>\n",
       "      <td>0.077578</td>\n",
       "      <td>b'gm12878'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sc12.ACTTGA</td>\n",
       "      <td>[nan, 0.007965973421810463, 0.0032579043423677...</td>\n",
       "      <td>0.046052</td>\n",
       "      <td>b'imr90'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>sc8.ACTTGA</td>\n",
       "      <td>[nan, 0.007160042006096354, 0.0189993520779955...</td>\n",
       "      <td>0.044051</td>\n",
       "      <td>b'imr90'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>sc9.ACTTGA</td>\n",
       "      <td>[nan, 0.00812826707197568, 0.01890383533408884...</td>\n",
       "      <td>0.044944</td>\n",
       "      <td>b'gm12878'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>sc9.CGATGT</td>\n",
       "      <td>[nan, 0.0078680376907839, -0.00099844023740305...</td>\n",
       "      <td>0.080890</td>\n",
       "      <td>b'imr90'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>sc9.GCCAAT</td>\n",
       "      <td>[nan, 0.013826760851188152, 0.0189999999975415...</td>\n",
       "      <td>0.116383</td>\n",
       "      <td>b'gm12878'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>sc9.TAGCTT</td>\n",
       "      <td>[nan, 0.01899079020752342, 0.019, 0.019, 0.018...</td>\n",
       "      <td>0.201357</td>\n",
       "      <td>b'gm12878'</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample                                    A/B Compartment  \\\n",
       "0    sc11.ACTTGA  [nan, 0.018999999977951314, 0.019, 0.019, 0.01...   \n",
       "1    sc11.CGATGT  [nan, -0.001, -0.0009999976557568116, 0.015259...   \n",
       "2    sc11.GCCAAT  [nan, -0.0009980968427463922, -0.0009999999999...   \n",
       "3    sc11.TAGCTT  [nan, 0.019, 0.018717003742121565, 0.019, 0.01...   \n",
       "4    sc12.ACTTGA  [nan, 0.007965973421810463, 0.0032579043423677...   \n",
       "..           ...                                                ...   \n",
       "124   sc8.ACTTGA  [nan, 0.007160042006096354, 0.0189993520779955...   \n",
       "125   sc9.ACTTGA  [nan, 0.00812826707197568, 0.01890383533408884...   \n",
       "126   sc9.CGATGT  [nan, 0.0078680376907839, -0.00099844023740305...   \n",
       "127   sc9.GCCAAT  [nan, 0.013826760851188152, 0.0189999999975415...   \n",
       "128   sc9.TAGCTT  [nan, 0.01899079020752342, 0.019, 0.019, 0.018...   \n",
       "\n",
       "     Correlation With Bulk   Cell Type  \n",
       "0                 0.115300  b'gm12878'  \n",
       "1                 0.169729    b'imr90'  \n",
       "2                 0.163800    b'imr90'  \n",
       "3                 0.077578  b'gm12878'  \n",
       "4                 0.046052    b'imr90'  \n",
       "..                     ...         ...  \n",
       "124               0.044051    b'imr90'  \n",
       "125               0.044944  b'gm12878'  \n",
       "126               0.080890    b'imr90'  \n",
       "127               0.116383  b'gm12878'  \n",
       "128               0.201357  b'gm12878'  \n",
       "\n",
       "[129 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosome_results['chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbafb91d",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "#compute the genome-wide correlation of the two AB compartment cell types\n",
    "# should be somehting like .6 if all the above steps were done correctly\n",
    "# this is a sanity check\n",
    "# There seems to be a problem here\n",
    "# I beleive the issue is that without KR normalization, the genome wide correlation of bulk is not that high\n",
    "########################################################################\n",
    "# Initialize lists to collect the eigenvector values for IMR90 and GM12878\n",
    "imr90_values = []\n",
    "gm12878_values = []\n",
    "\n",
    "# Loop through each chromosome\n",
    "for i in range(1, 23):\n",
    "    key_gm12878 = f'res{resolution}_ch{i}_oe_GM12878_{normalization}_eigenvector'\n",
    "    key_imr90 = f'res{resolution}_ch{i}_oe_IMR90_{normalization}_eigenvector'                                                               \n",
    "                                                               \n",
    "    # Check if the keys exist in the data to avoid KeyErrors\n",
    "    if key_gm12878 in bulk_data and key_imr90 in bulk_data:\n",
    "        # Extend the list with the eigenvector values from each chromosome\n",
    "        # Ensure values are numeric (floats or ints)\n",
    "        gm12878_vals = bulk_data[key_gm12878]['eigenvalue']\n",
    "        imr90_vals = bulk_data[key_imr90]['eigenvalue']\n",
    "        imr90_values.extend(imr90_vals)\n",
    "        gm12878_values.extend(gm12878_vals)\n",
    "        \n",
    "# Convert lists to pandas Series for correlation calculation\n",
    "imr90_series = pd.Series(imr90_values)\n",
    "gm12878_series = pd.Series(gm12878_values)\n",
    "\n",
    "# Calculate the Pearson correlation coefficient between IMR90 and GM12878\n",
    "correlation = imr90_series.corr(gm12878_series)\n",
    "\n",
    "print(f'Genome-wide correlation between IMR90 and GM12878: {correlation}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a80a1ebe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef8a5868",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90585725",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d075bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chromosome_results['chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dd62604",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14c89a3d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51c8f7e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c36c9f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This block needs to make a caption and save an image\n",
    "#sort and make single chromoosme image\n",
    "\n",
    "gm12878_count = sc_dataframe[sc_dataframe['Cell Type'] == 'gm12878'].shape[0]\n",
    "gm12878_count\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "chr=1    \n",
    "# Assuming chromosome_results['chr1'] contains relevant data\n",
    "sc_dataframe = chromosome_results[f'chr{chr}']\n",
    "\n",
    "# First, sort by 'Cell Type' in a way that 'gm12878' cells come first, and then by 'Correlation With Bulk' descending\n",
    "sc_dataframe['Cell Type Sort'] = sc_dataframe['Cell Type'].map({'gm12878': 0, 'imr90': 1})  # Helps in sorting by cell type\n",
    "sc_dataframe = sc_dataframe.sort_values(by=['Cell Type Sort', 'Correlation With Bulk'], ascending=[True, False])\n",
    "\n",
    "# Remove the auxiliary column after sorting\n",
    "sc_dataframe.drop('Cell Type Sort', axis=1, inplace=True)\n",
    "\n",
    "# Now sc_dataframe is sorted first by cell type ('gm12878' first) and then by 'Correlation With Bulk' within each type\n",
    "sc_dataframe\n",
    "\n",
    "\n",
    "def plot_eigenvectors_with_heatmap(bulk_eigenvector1, bulk_eigenvector2, sc_dataframe, title1, title2, title, save_path=None, separation_rows=5):\n",
    "    fig = plt.figure(figsize=(15, 20))\n",
    "    gs = gridspec.GridSpec(nrows=3, ncols=2, figure=fig, width_ratios=[30, 1], height_ratios=[1, 1, 4], hspace=0.3, wspace=0.05)\n",
    "\n",
    "    # Plot Bulk Eigenvectors\n",
    "    ax_bulk1 = fig.add_subplot(gs[0, 0])\n",
    "    ax_bulk1.plot(bulk_eigenvector1, color='blue')\n",
    "    ax_bulk1.set_title(title1, fontsize=25)\n",
    "    ax_bulk1.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_bulk1.set_ylabel(\"Eigenvector Value\", fontsize=20)\n",
    "\n",
    "    ax_bulk2 = fig.add_subplot(gs[1, 0])\n",
    "    ax_bulk2.plot(bulk_eigenvector2, color='green')\n",
    "    ax_bulk2.set_title(title2, fontsize=25)\n",
    "    ax_bulk2.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_bulk2.set_ylabel(\"Eigenvector Value\", fontsize=20)\n",
    "\n",
    "    # Add separation rows\n",
    "    gm12878_data = np.stack(sc_dataframe[sc_dataframe['Cell Type'] == 'gm12878']['A/B Compartment'].apply(lambda x: np.array(x)).tolist())\n",
    "    imr90_data = np.stack(sc_dataframe[sc_dataframe['Cell Type'] == 'imr90']['A/B Compartment'].apply(lambda x: np.array(x)).tolist())\n",
    "    \n",
    "    # Create empty rows filled with NaNs for separation\n",
    "    empty_rows = np.full((separation_rows, gm12878_data.shape[1]), np.nan)\n",
    "    \n",
    "    # Stack GM12878 data, empty rows, and IMR90 data\n",
    "    sc_data_array = np.vstack([gm12878_data, empty_rows, imr90_data])\n",
    "\n",
    "    # Create the heatmap with a red colormap\n",
    "    ax_sc = fig.add_subplot(gs[2, 0])\n",
    "    im = ax_sc.imshow(sc_data_array, aspect='auto', cmap='Reds', interpolation='nearest')\n",
    "    ax_sc.set_title(\"Heatmap of Single Cell Samples\", fontsize=20)\n",
    "    ax_sc.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_sc.set_ylabel(\"Sample Index\", fontsize=20)\n",
    "\n",
    "    # Add labels for GM12878 and IMR90 cell types\n",
    "    num_gm12878 = len(gm12878_data)\n",
    "    num_imr90 = len(imr90_data)\n",
    "    \n",
    "    ax_sc.set_yticks([num_gm12878 / 2, num_gm12878 + separation_rows + num_imr90 / 2])\n",
    "    ax_sc.set_yticklabels(['GM12878', 'IMR90'], fontsize=30)\n",
    "\n",
    "    # Colorbar for the heatmap\n",
    "    ax_cbar = fig.add_subplot(gs[2, 1])\n",
    "    fig.colorbar(im, cax=ax_cbar)\n",
    "    ax_cbar.set_ylabel('Compartment Call Value', rotation=270, labelpad=25, fontsize=15)\n",
    "\n",
    "    plt.suptitle(title, fontsize=30)\n",
    "\n",
    "    # Save the figure if a save path is provided\n",
    "    if save_path:\n",
    "        plt.savefig(save_path, bbox_inches='tight')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "# Example invocation\n",
    "plot_eigenvectors_with_heatmap(\n",
    "    bulk_eigenvector_gm12878,\n",
    "    bulk_eigenvector_imr90,\n",
    "    sc_dataframe,\n",
    "    \"Bulk for chr1 - GM12878\",\n",
    "    \"Bulk for chr1 - IMR90\",\n",
    "    \"Single Cell Compartment Call Comparison\",\n",
    "    save_path=\"../../files/AB_compartment_heatmap.png\"  # Specify the file path to save the image\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aa3121d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2930f051",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bef0c359",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f90a5419",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35fb8acd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "from config_and_print import methy_directory, filtered_list, chrom_file, resolutions, output_directory, mappability_threshold\n",
    "#chromosomes = [f'chr{chrom}' for chrom in chromosomes]\n",
    "\n",
    "# Ensure resolutions is treated as a tuple or list of strings\n",
    "if isinstance(resolutions, str):\n",
    "    resolutions = (resolutions,)\n",
    "\n",
    "# Print resolutions for debugging\n",
    "print(f\"Resolutions from config: {resolutions}\")\n",
    "\n",
    "# Extract resolution value and label from the resolutions string\n",
    "resolution_str = resolutions[0]\n",
    "\n",
    "# Debug print to check the value of resolution_str\n",
    "print(f\"Extracted resolution string: {resolution_str}\")\n",
    "\n",
    "def parse_resolution(resolution_str):\n",
    "    if ':' in resolution_str:\n",
    "        resolution_value, resolution_label = resolution_str.split(':')\n",
    "        try:\n",
    "            resolution = int(resolution_value)\n",
    "            return resolution, resolution_label\n",
    "        except ValueError:\n",
    "            raise ValueError(f\"Resolution value should be an integer: '{resolution_value}' in '{resolution_str}'\")\n",
    "    else:\n",
    "        raise ValueError(f\"Invalid resolution format: '{resolution_str}'. Expected format 'value:label', e.g., '1000000:1Mb'.\")\n",
    "\n",
    "resolution, resolution_label = parse_resolution(resolution_str)\n",
    "\n",
    "# Define the bulk eigenvectors\n",
    "bulk_eigenvector_gm12878 = original_bulk_data[f'res1000000_ch1_oe_GM12878_{normalization}_eigenvector']['eigenvalue'].values\n",
    "bulk_eigenvector_imr90 = original_bulk_data[f'res1000000_ch1_oe_IMR90_{normalization}_eigenvector']['eigenvalue'].values\n",
    "\n",
    "# Generate the plot\n",
    "chr=1    \n",
    "sc_dataframe = chromosome_results[f'chr{chr}']\n",
    "\n",
    "# Sort and prepare the dataframe\n",
    "sc_dataframe['Cell Type Sort'] = sc_dataframe['Cell Type'].map({'gm12878': 0, 'imr90': 1})\n",
    "sc_dataframe = sc_dataframe.sort_values(by=['Cell Type Sort', 'Correlation With Bulk'], ascending=[True, False])\n",
    "sc_dataframe.drop('Cell Type Sort', axis=1, inplace=True)\n",
    "\n",
    "# Function to plot eigenvectors and heatmap\n",
    "def plot_eigenvectors_with_heatmap(bulk_eigenvector1, bulk_eigenvector2, sc_dataframe, title1, title2, title, save_path=None, separation_rows=5):\n",
    "    fig = plt.figure(figsize=(15, 20))\n",
    "    gs = gridspec.GridSpec(nrows=3, ncols=2, figure=fig, width_ratios=[30, 1], height_ratios=[1, 1, 4], hspace=0.3, wspace=0.05)\n",
    "\n",
    "    ax_bulk1 = fig.add_subplot(gs[0, 0])\n",
    "    ax_bulk1.plot(bulk_eigenvector1, color='blue')\n",
    "    ax_bulk1.set_title(title1, fontsize=25)\n",
    "    ax_bulk1.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_bulk1.set_ylabel(\"Eigenvector Value\", fontsize=20)\n",
    "\n",
    "    ax_bulk2 = fig.add_subplot(gs[1, 0])\n",
    "    ax_bulk2.plot(bulk_eigenvector2, color='green')\n",
    "    ax_bulk2.set_title(title2, fontsize=25)\n",
    "    ax_bulk2.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_bulk2.set_ylabel(\"Eigenvector Value\", fontsize=20)\n",
    "\n",
    "    gm12878_data = np.stack(sc_dataframe[sc_dataframe['Cell Type'] == 'gm12878']['A/B Compartment'].apply(lambda x: np.array(x)).tolist())\n",
    "    imr90_data = np.stack(sc_dataframe[sc_dataframe['Cell Type'] == 'imr90']['A/B Compartment'].apply(lambda x: np.array(x)).tolist())\n",
    "    \n",
    "    empty_rows = np.full((separation_rows, gm12878_data.shape[1]), np.nan)\n",
    "    sc_data_array = np.vstack([gm12878_data, empty_rows, imr90_data])\n",
    "\n",
    "    ax_sc = fig.add_subplot(gs[2, 0])\n",
    "    im = ax_sc.imshow(sc_data_array, aspect='auto', cmap='Reds', interpolation='nearest')\n",
    "    ax_sc.set_title(\"Heatmap of Single Cell Samples\", fontsize=20)\n",
    "    ax_sc.set_xlabel(\"Bin Index\", fontsize=20)\n",
    "    ax_sc.set_ylabel(\"Sample Index\", fontsize=20)\n",
    "\n",
    "    num_gm12878 = len(gm12878_data)\n",
    "    num_imr90 = len(imr90_data)\n",
    "    \n",
    "    ax_sc.set_yticks([num_gm12878 / 2, num_gm12878 + separation_rows + num_imr90 / 2])\n",
    "    ax_sc.set_yticklabels(['GM12878', 'IMR90'], fontsize=30)\n",
    "\n",
    "    ax_cbar = fig.add_subplot(gs[2, 1])\n",
    "    fig.colorbar(im, cax=ax_cbar)\n",
    "    ax_cbar.set_ylabel('Compartment Call Value', rotation=270, labelpad=25, fontsize=15)\n",
    "\n",
    "    plt.suptitle(title, fontsize=30)\n",
    "\n",
    "    if save_path:\n",
    "        plt.savefig(save_path, bbox_inches='tight')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "# Example invocation\n",
    "plot_eigenvectors_with_heatmap(\n",
    "    bulk_eigenvector_gm12878,\n",
    "    bulk_eigenvector_imr90,\n",
    "    sc_dataframe,\n",
    "    f\"Bulk for chr{chr} - GM12878\",\n",
    "    f\"Bulk for chr{chr} - IMR90\",\n",
    "    \"Single Cell Compartment Call Comparison\",\n",
    "    save_path=\"../../../files/AB_compartment_heatmap.png\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff458b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome_results['chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9217f954",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiomics6",
   "language": "python",
   "name": "multiomics6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
