{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "976ebe7f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bdca2639",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing /home/dwk681/workspace/cluster_cells_from_GSE189158_NOMe_HiC/filesFromCluster/bam/methylation/filter_low_qual/sc24.ACTTGA.methy.b37.bw\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr1/sc24.ACTTGA_chr1.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr2/sc24.ACTTGA_chr2.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr3/sc24.ACTTGA_chr3.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr4/sc24.ACTTGA_chr4.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr5/sc24.ACTTGA_chr5.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr6/sc24.ACTTGA_chr6.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr7/sc24.ACTTGA_chr7.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr8/sc24.ACTTGA_chr8.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr9/sc24.ACTTGA_chr9.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr10/sc24.ACTTGA_chr10.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr11/sc24.ACTTGA_chr11.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr12/sc24.ACTTGA_chr12.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr13/sc24.ACTTGA_chr13.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr14/sc24.ACTTGA_chr14.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr15/sc24.ACTTGA_chr15.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr16/sc24.ACTTGA_chr16.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr17/sc24.ACTTGA_chr17.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr18/sc24.ACTTGA_chr18.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr19/sc24.ACTTGA_chr19.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr20/sc24.ACTTGA_chr20.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr21/sc24.ACTTGA_chr21.h5 already exists, skipping computation.\n",
      "File ../../projects/single_cell_files/methy_1Mb_correlation_dir/chr22/sc24.ACTTGA_chr22.h5 already exists, skipping computation.\n",
      "Processing /home/dwk681/workspace/cluster_cells_from_GSE189158_NOMe_HiC/filesFromCluster/bam/methylation/filter_low_qual/sc14.TAGCTT.methy.b37.bw\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 115\u001b[0m\n\u001b[1;32m    113\u001b[0m     large_bin_size \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mint\u001b[39m(res_value)\n\u001b[1;32m    114\u001b[0m     small_bin_size \u001b[38;5;241m=\u001b[39m large_bin_size \u001b[38;5;241m/\u001b[39m\u001b[38;5;241m/\u001b[39m \u001b[38;5;241m100\u001b[39m\n\u001b[0;32m--> 115\u001b[0m     \u001b[43mprocess_bigwig_file\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbigwigfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlarge_bin_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msmall_bin_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mres_label\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbase_directory\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefix\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    116\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m:\n\u001b[1;32m    117\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInvalid resolution format: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresolution\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[10], line 65\u001b[0m, in \u001b[0;36mprocess_bigwig_file\u001b[0;34m(bigwigfile, large_bin_size, small_bin_size, resolution_label, base_directory, prefix)\u001b[0m\n\u001b[1;32m     62\u001b[0m small_bin_data \u001b[38;5;241m=\u001b[39m bw\u001b[38;5;241m.\u001b[39mvalues(chrom, \u001b[38;5;28mint\u001b[39m(small_bin_start), \u001b[38;5;28mint\u001b[39m(small_bin_end))\n\u001b[1;32m     64\u001b[0m \u001b[38;5;66;03m# Calculate the integral of data in the small bin\u001b[39;00m\n\u001b[0;32m---> 65\u001b[0m integral \u001b[38;5;241m=\u001b[39m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrapz\u001b[49m\u001b[43m(\u001b[49m\u001b[43msmall_bin_data\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     67\u001b[0m \u001b[38;5;66;03m# Store the integral in the corresponding list\u001b[39;00m\n\u001b[1;32m     68\u001b[0m integrals[j, i] \u001b[38;5;241m=\u001b[39m integral\n",
      "File \u001b[0;32m<__array_function__ internals>:200\u001b[0m, in \u001b[0;36mtrapz\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "File \u001b[0;32m~/.conda/envs/multiomics6/lib/python3.8/site-packages/numpy/lib/function_base.py:4859\u001b[0m, in \u001b[0;36mtrapz\u001b[0;34m(y, x, dx, axis)\u001b[0m\n\u001b[1;32m   4773\u001b[0m \u001b[38;5;129m@array_function_dispatch\u001b[39m(_trapz_dispatcher)\n\u001b[1;32m   4774\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mtrapz\u001b[39m(y, x\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, dx\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1.0\u001b[39m, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m):\n\u001b[1;32m   4775\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4776\u001b[0m \u001b[38;5;124;03m    Integrate along the given axis using the composite trapezoidal rule.\u001b[39;00m\n\u001b[1;32m   4777\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4857\u001b[0m \u001b[38;5;124;03m    array([2.,  8.])\u001b[39;00m\n\u001b[1;32m   4858\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 4859\u001b[0m     y \u001b[38;5;241m=\u001b[39m \u001b[43masanyarray\u001b[49m\u001b[43m(\u001b[49m\u001b[43my\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4860\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m x \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   4861\u001b[0m         d \u001b[38;5;241m=\u001b[39m dx\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pyBigWig\n",
    "import numpy as np\n",
    "from scipy.stats import ks_2samp\n",
    "import h5py\n",
    "\n",
    "# Import variables from config_and_print.py\n",
    "from config_and_print import methy_directory, output_directory, resolutions, filtered_list\n",
    "\n",
    "# Read the filtered bam list\n",
    "with open(filtered_list, 'r') as f:\n",
    "    filtered_prefixes = set(line.strip() for line in f)\n",
    "\n",
    "# Function to process each bigWig file\n",
    "def process_bigwig_file(bigwigfile, large_bin_size, small_bin_size, resolution_label, base_directory, prefix):\n",
    "    bw = pyBigWig.open(bigwigfile)\n",
    "    chromosomes = bw.chroms()\n",
    "\n",
    "    for chrom in chromosomes:\n",
    "        # Ensure chrom_string is always in the format 'chrN'\n",
    "        if isinstance(chrom, int) or chrom.isdigit():\n",
    "            chrom_string = f'chr{chrom}'\n",
    "        else:\n",
    "            chrom_string = chrom if chrom.startswith('chr') else f'chr{chrom}'\n",
    "\n",
    "        if chrom_string in [f'chr{i}' for i in range(1, 23)]:\n",
    "            # Create chromosome-specific directory if it doesn't exist\n",
    "            chrom_directory = os.path.join(base_directory, chrom_string)\n",
    "            os.makedirs(chrom_directory, exist_ok=True)\n",
    "\n",
    "            # Define the filename for the matrix\n",
    "            new_filename = os.path.join(chrom_directory, f'{prefix}_{chrom_string}.h5')\n",
    "\n",
    "            # Check if the file already exists\n",
    "            if os.path.exists(new_filename):\n",
    "                print(f\"File {new_filename} already exists, skipping computation.\")\n",
    "                continue\n",
    "\n",
    "            chrom_length = chromosomes[chrom]\n",
    "            large_bins = range(0, chrom_length, large_bin_size)\n",
    "            num_large_bins = len(large_bins)\n",
    "\n",
    "            integrals = np.zeros((large_bin_size // small_bin_size, num_large_bins))\n",
    "\n",
    "            for i, large_bin_start in enumerate(large_bins):\n",
    "                for j in range(large_bin_size // small_bin_size):\n",
    "                    small_bin_start = large_bin_start + j * small_bin_size\n",
    "                    small_bin_end = small_bin_start + small_bin_size\n",
    "\n",
    "                    # Ensure that the coordinates are within the chromosome boundaries\n",
    "                    if small_bin_start < 0:\n",
    "                        small_bin_start = 0\n",
    "                    if small_bin_end > chrom_length:\n",
    "                        small_bin_end = chrom_length\n",
    "\n",
    "                    # Check if the adjusted small_bin_start is still greater than or equal to small_bin_end\n",
    "                    if small_bin_start >= small_bin_end:\n",
    "                        continue\n",
    "\n",
    "                    try:\n",
    "                        # Get data within the small bin\n",
    "                        small_bin_data = bw.values(chrom, int(small_bin_start), int(small_bin_end))\n",
    "\n",
    "                        # Calculate the integral of data in the small bin\n",
    "                        integral = np.trapz(small_bin_data)\n",
    "\n",
    "                        # Store the integral in the corresponding list\n",
    "                        integrals[j, i] = integral\n",
    "                    except Exception as e:\n",
    "                        # Print the exception and relevant information\n",
    "                        print(f\"An error occurred for chromosome {chrom}, large bin {i}, small bin {j}: {e}\")\n",
    "\n",
    "            # create your matrices\n",
    "            num_rows, num_columns = integrals.shape\n",
    "            p_value_ks_matrix = np.zeros((num_columns, num_columns))\n",
    "            epsilon = 1e-10  # Define your epsilon value here\n",
    "\n",
    "            for i in range(num_columns):\n",
    "                for j in range(i, num_columns):\n",
    "                    if i == j:\n",
    "                        p_value_ks_matrix[i, j] = 1.0\n",
    "                    else:\n",
    "                        _, ks_p_value = ks_2samp(integrals[:, i], integrals[:, j])\n",
    "\n",
    "                        p_value_ks_matrix[i, j] = np.log10(ks_p_value + epsilon)\n",
    "                        p_value_ks_matrix[j, i] = np.log10(ks_p_value + epsilon)\n",
    "\n",
    "            # take the correlation matrices of all the matrices\n",
    "            correlation_ks_p_value = np.corrcoef(p_value_ks_matrix) + 1\n",
    "\n",
    "            # Save the correlation epigenetic matrix as a .h5 file\n",
    "            with h5py.File(new_filename, 'w') as hf:\n",
    "                hf.create_dataset('methy_matrix', data=correlation_ks_p_value)\n",
    "\n",
    "# Ensure resolutions is processed as a list of strings\n",
    "resolutions_list = [res.strip() for res in resolutions.split(',')]\n",
    "\n",
    "# Iterate through the files in methy_directory\n",
    "for root, dirs, files in os.walk(methy_directory):\n",
    "    for file in files:\n",
    "        if file.endswith('.methy.b37.bw'):\n",
    "            # Extract the prefix for filtering\n",
    "            prefix = '.'.join(file.split('.')[:2])\n",
    "            if prefix in filtered_prefixes:\n",
    "                bigwigfile = os.path.join(root, file)\n",
    "                print(f\"Processing {bigwigfile}\")\n",
    "                for resolution in resolutions_list:\n",
    "                    try:\n",
    "                        res_value, res_label = resolution.split(\":\")\n",
    "                        # Create the base directory for the matrices\n",
    "                        base_directory = os.path.join(output_directory, f'methy_{res_label}_correlation_dir')\n",
    "                        os.makedirs(base_directory, exist_ok=True)\n",
    "                        large_bin_size = int(res_value)\n",
    "                        small_bin_size = large_bin_size // 100\n",
    "                        process_bigwig_file(bigwigfile, large_bin_size, small_bin_size, res_label, base_directory, prefix)\n",
    "                    except ValueError:\n",
    "                        print(f\"Invalid resolution format: {resolution}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4c47dd9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1000000', '1Mb']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ade8fd8d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiomics6",
   "language": "python",
   "name": "multiomics6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
